chain/consensus/hierarchical/actors/sca/sca_apply.go

```diff
@@ -41,6 +52,67 @@ func applyFund(rt runtime.Runtime, msg types.Message) {
 	}
 }
 
+func applyRelease(rt runtime.Runtime, msg types.Message) {
+	var st SCAState
+
+	rfrom, rto := fromToRawAddr(rt, msg.From, msg.To)
+	snFrom, err := msg.From.Subnet()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error getting subnet from HAddress")
+
+	// Check that msg.From == BurntRewardsActorAddr, this determines that is a release message
+	if rfrom != builtin.BurntFundsActorAddr {
+		rt.Abortf(exitcode.ErrIllegalArgument, "msg passed as argument not a funding message (from != burtnRewardActorAddr)")
+	}
+
+	rt.StateTransaction(&st, func() {
+		bottomUpStateTransition(rt, &st, msg)
+
+		// Update circulating supply reducing release value.
+		sh, has, err := st.GetSubnet(adt.AsStore(rt), snFrom)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error fetching subnet state")
+		if !has {
+			rt.Abortf(exitcode.ErrIllegalState, "subnet for actor hasn't been registered yet")
+		}
+		// Even if the actor has balance, we shouldn't allow releasing more than
+		// the current circulating supply, as it would mean that we are
+		// releasing funds from the collateral.
+		if sh.CircSupply.LessThan(msg.Value) {
+			rt.Abortf(exitcode.ErrIllegalState, "wtf! we can't release funds below the circulating supply. Something went wrong!")
+		}
+		sh.CircSupply = big.Sub(sh.CircSupply, msg.Value)
+		st.flushSubnet(rt, sh)
+	})
+
+	// Release funds to the destination address.
+	code := rt.Send(rto, builtin.MethodSend, nil, msg.Value, &builtin.Discard{})
+	if !code.IsSuccess() {
+		noop(rt, code)
+	}
+}
+
+func bottomUpStateTransition(rt runtime.Runtime, st *SCAState, msg types.Message) {
+	// Bottom-up messages include the nonce of their message meta. Several messages
+	// will include the same nonce. They need to be applied in order of nonce.
+
+	// As soon as we see a message with the next msgMeta nonce, we increment the nonce
+	// and start accepting the one for the next nonce.
+	// FIXME: Once we have the end-to-end flow of cross-message this REALLY needs to
+	// be revisited.
+	if st.AppliedBottomUpNonce+1 == msg.Nonce {
+		// Increment latest nonce applied for bottomup
+		incrementNonce(rt, &st.AppliedBottomUpNonce)
+	}
+
+	// NOTE: Check if the nonce of the message being applied is the subsequent one (we could relax a bit this
+	// requirement, but it would mean that we need to determine how we want to handle gaps, and messages
+	// being validated out-of-order).
+	if st.AppliedBottomUpNonce != msg.Nonce {
+		rt.Abortf(exitcode.ErrIllegalState, "the message being applied doesn't hold the subsequent nonce (nonce=%d, applied=%d)",
```

Comment:
What happens in case of Abort? Can i get back my burned funds?

Reply:
or retry with a higher nonce

Reply:
That is a great question. The answer in the current implementation is _no_. This is something that we need to figure out how we want to do it. We could maybe add some information in checkpoint to notify back the source that certain message could not be applied (although this may be quite slow). I think we briefly mentioned this in our previous sync, but let me add a point to discuss it again in our next sync.